crates/analyzer/src/traversal/expressions.rs

```diff
@@ -623,7 +623,14 @@ fn expr_bin_operation(
     expected_type: Option<Integer>,
 ) -> Result<ExpressionAttributes, FatalError> {
     if let fe::Expr::BinOperation { left, op, right } = &exp.kind {
-        let expected = expected_type.map(Type::int);
+        let expected = match &op.kind {
+            // In shift operations, the right hand side may have a different type than the left hand side
+            // because the right hand side needs to be unsigned. The type of the entire expression is
+            // determined by the left hand side anyway so we don't try to coerce the right hand side in this case.
+            fe::BinOperator::LShift | fe::BinOperator::RShift => None,
+            _ => expected_type.map(Type::int),
+        };
+
         let left_attributes = value_expr(scope, left, expected.as_ref())?;
         let right_attributes = value_expr(scope, right, expected.as_ref())?;
```

Comment:
I agree that the RHS shouldn't be coerced, but shouldn't the LHS be? With this change:
```
error: type mismatch
  ┌─ /home/sean/code/eth/shift.fe:4:18
  │
4 │     let x: i64 = -1 << 0
  │                  ^^^^^^^ this has type `i256`; expected type `i64`
```

Reply:
Mmh, that's right (side note, it doesn't work on master either) but I wonder if it is safe to simply coerce the return type into `i64` or if an actual type cast would be needed.

When you do `let foo: i8 = -1` the generated yul is `let $foo := sub(0, 1)` but when you do `let foo: i8 = -1` the generated yul is  `let $foo := adjust_numeric_i8(sub(0, 1))`. *Maybe*[1] the call to `adjust_numeric_*` isn't needed when we are dealing with a literal because the range of the literal is checked at compile time but I would expect it is important for all other cases. So, I guess what we really want is `let x: i64 = i64(-1 << 0)` which **does work** and produces the following yul: `let $x := adjust_numeric_i64(shl(0, sub(0, 1)))`

I have a hard time to decide what to do here:

- should we just coerce the return type of the binary operation to the expected type?
  - if yes then in all cases or just within boundaries?
- should we continue to error and have the user add a manual type cast
- should we use lowering to add type casts (and maybe even replace the current coercing mechanic which sidesteps the numeric adjustment calls)

As you can see, I'm desperately in need of a professional to help me out here :joy_cat: 
Also paging in @g-r-a-n-t because he already spend some time thinking about the cleanup stuff, too. 

[1] I have a hard time telling if that is actually true. The more I think about it, the less I know.

Reply:
I think in the case of values that are known at compile-time, we should error if the result of the operation overflows the target type, and simply coerce into the target type if it doesn't. Ideally, we'd just emit the resulting value rather than a call to `shl`(. For now, "known at compile-time" will just be literals, but someday the compiler will be sufficiently smart to propagate constant values from previous lines of code and could detect overflow on something like `x << y` if x and y have known values at compile-time.)

If the values *aren't* known at compile time:

1) if the type of the LHS matches the target type: check for overflow at runtime and revert?
2) if the type of the LHS doesn't match the target type: require a manual type cast

For `let foo: i8 = i8(-1)`, I think if the value is a constant, this should be exactly the same as `let foo: i8 = -1`, and ideally the yul would be `let $foo := 0x11111...` (though maybe the yul optimizer would replace `sub(0, 1)` with a constant for us anyway). It seems to me that we don't need or want `adjust_numeric` calls when the value is known to fit within the target type, right? The value will already be sign-extended.

Reply:
Thanks for your answer. So, I agree with everything you wrote about dealing with values that are known at compile time, that eventually we should be able to map them into a literal.

So, let's focus on the case where they *aren't* known at compile time:

>if the type of the LHS matches the target type: check for overflow at runtime and revert

This is exactly what we do right now. Also note that the overflow check is currently always based on the type of the LHS.

>if the type of the LHS doesn't match the target type: require a manual type cast

Yes, that's why I was thinking that `let x: i64 = -1 << 0` is ok to error because `-1` is of type `i256` and so the user needs to manually cast it: `let x: i64 = i64(-1 << 0)`. Because when I was looking at ``let x: i64 = -1 << 0` what I was actually thinking about was ``let x: i64 = some_i256 << 0` and not a literal. But now I realized that the type coercion is only applied to a literal so that the following continues to be an error (rightfully so I think)

```
contract Foo:

  pub fn bar(z: i256):
    let x: i64 = z << 0
```

But this has revealed to me that I don't really understand how the type coercion based on passing `expected_type` works. How is it that the compiler doesn't try to coerce `z` into an `i64`? I tried to answer that myself but it isn't obvious to me how that works :grimacing: 

Apart from my open question, I believe the PR is now doing the right thing, doesn't it?

Reply:
> How is it that the coercion is only applied to literals and the compiler doesn't try to coerce z into an i64?

The expected type is only used when the type of an expression isn't already known (we consider `10` to be of some unknown integer type, and `[]` to be some unknown array type, etc). `expr_name` doesn't use the expected type because the types of all variables are explicitly declared; it returns the declared type and lets the caller decide if the types are compatible or not.

> Apart from my open question, I believe the PR is now doing the right thing, doesn't it?

Yep, looks good to me. As far as I can tell, we aren't yet evaluating operations and checking for overflow on the resulting value (for constants, at compile time I mean). Eg `let x: u8 = 1 << 9` could/should be an analyzer error, but isn't. A job for a future PR though.

Reply:
Ah, that's how the magic works. Thanks!

>checking for overflow

That's true but afaik Solidity isn't doing that either (Rust does though afaik). I *think* it's intentional and that one can look at it from two different angles.

Let's consider the value `1` as an `u8` which is represented as
the following 256 bit item on the EVM stack `00..|00000001|`. A left shift of `8` bits (`val << 8`) turns that into `00..01|00000000|`. We can say that this overflows but we could also simply consider it all zeros now since the `1` has moved out of the 8 bit range. We do cleanup the value so that it eventually turns into `00..00|00000000|` then.

Reply:
I think we should definitely call that an error at compile time, and perhaps suggest `let x: u8 = u8(1 << 8)` as a way to explicitly truncate and silence the error.